### DISCUSSION TITLE
Prisma: Converting prisma 1 scalar/enum to prisma 2 introspection issue

### DISCUSSION THREAD

**Discussion by TheoMer** [prisma/prisma#3058] [Q&A] [answered]
```
"prisma1": "^1.34.11",
"ts-node-dev": "^1.0.0-pre.50"
"ts-node": "^8.10.2",
"nexus": "^0.26.0-next.3",
"nexus-plugin-prisma": "^0.16.1"
```

So I'm in the process of migrating from Prisma 1 to Prisma 2 and don't know how to correctly handle the following in PostgreSQL so that the model is correctly reflected after introspection:

**Prisma 1 datamodel**

```graphql
enum Permission {
  ADMIN
  USER
  GUEST_USER
  ITEMCREATE
  ITEMUPDATE
  ITEMDELETE
  PERMISSIONUPDATE
}

type User {
  permissions: [Permission] @scalarList(strategy: RELATION)
}
```

**Prisma 1 generated SQL migration**

```sql
CREATE TABLE "flam-production$prod"."User_permissions" (
	"nodeId" varchar(25) NOT NULL,
	"position" int4 NOT NULL,
	value text NOT NULL,
	CONSTRAINT "User_permissions_pkey" PRIMARY KEY ("nodeId", "position")
);


-- "flam-production$prod"."User_permissions" foreign keys

ALTER TABLE "flam-production$prod"."User_permissions" ADD CONSTRAINT "User_permissions_nodeId_fkey" FOREIGN KEY ("nodeId") REFERENCES "flam-production$prod"."User"(id);
```

**Result of Prisma 2.0 introspection**

```prisma
model User {
  User_permissions User_permissions[]
}

model User_permissions {
  nodeId   String
  position Int
  value    String
  User     User   @relation(fields: [nodeId], references: [id])

  @@id([nodeId, position])
```

Is there a workaround I can use in postgresql to ensure that introspection produces something similar to the following schema.prisma output?

```prisma
enum Permission {
  ADMIN
  USER
  GUEST_USER
  ITEMCREATE
  ITEMUPDATE
  ITEMDELETE
  PERMISSIONUPDATE
}

model User {
  permissions Permission[]
}
```

**Comment by nikolasburk** [comment, marked as answer]
Hey @TheoMer, thanks a lot for raising this! Are you using the [Upgrade CLI](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/how-to-upgrade#prisma-upgrade-cli) for the upgrade process?

The issue you're seeing is a mix of two [schema incompatibilities](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/how-to-upgrade#schema-incompatibilities):


- [Enums are represented as `TEXT` in database](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities/#enums-are-represented-as-text-in-database)
- [Scalar lists (arrays) are maintained with extra table](https://www.prisma.io/docs/guides/upgrade-guides/upgrade-from-prisma-1/schema-incompatibilities/#scalar-lists-arrays-are-maintained-with-extra-table)

Since enums _are_ scalars, you're experiencing both issues at the same time. Nevertheless, the Upgrade CLI should cover this case by generating the right SQL statements for you such that your end result of the Prisma schema would be exactly the one you describe:

```prisma
enum Permission {
  ADMIN
  USER
  GUEST_USER
  ITEMCREATE
  ITEMUPDATE
  ITEMDELETE
  PERMISSIONUPDATE
}

model User {
  permissions Permission[]
}
```

Would you mind "moving" this discussion as an [issue](https://github.com/prisma/upgrade/issues/new) in the `upgrade` repo? We'll pick it up there and ensure the use case is being addressed with the Upgrade CLI! 🙂

**Reply by TheoMer** [reply]
Hey @nikolasburk Many thanks for informing me about the upgrade cli, I was completely unaware of it. 

I will move this over to the upgrade repo as you suggested.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that have already been marked as answered but remain open.

If this discussion still requires further input or clarification, feel free to reopen it or start a new one with updated details. Your contributions are invaluable to the community, and we’re here to help!

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and ongoing support of the Prisma community!
